### PR TITLE
🌱 Relocate version change validation to preflight check 

### PR DESCRIPTION
**Pull request by Fedosin**
**What this PR does / why we need it**:

Previously, the check was executed at the end of reconciliation and in multiple instances. By moving it to the beginning, the process is halted when installed and desired versions match, streamlining the reconciliation workflow.

### CODE REVIEW

**Comment by Fedosin** [comment]
The issue is addressed in https://github.com/kubernetes-sigs/cluster-api-operator/pull/168, so the only meaningful thing we can take from this PR is to move preflight checks into a separate package.

File: internal/controller/genericprovider_controller.go [hunk 1]
```diff
@@ -127,7 +140,6 @@ func patchProvider(ctx context.Context, provider genericprovider.GenericProvider
 func (r *GenericProviderReconciler) reconcile(ctx context.Context, provider genericprovider.GenericProvider, genericProviderList genericprovider.GenericProviderList) (ctrl.Result, error) {
 	reconciler := newPhaseReconciler(*r, provider, genericProviderList)
 	phases := []reconcilePhaseFn{
-		reconciler.preflightChecks,
```
Comment (damdo) [inline internal/controller/genericprovider_controller.go:130]: What if we kept the preflightChecks as a phase here so we can benefit from the centralised condition setting logic, and rather we defined the `halt` pattern you came up with, as a pattern for all the phases.

Maybe with a customisable halting message (similar to the one you specified [here](https://github.dev/Fedosin/cluster-api-operator/blob/b17b9c91be17e774101047058762226f192bf96c/internal/controller/genericprovider_controller.go#L117-L118)) that can be logged before returning reconcile()? We could even have `halt` as a message string ()`haltMsg`?), and when the phase returns a non empty msg then we log and halt?

So [here](https://github.com/Fedosin/cluster-api-operator/blob/5c223b751ebc4c19fc4bdcc0ba8027ecbbd56fdb/internal/controller/genericprovider_controller.go) doing something like
```
res, halt, err = phase(ctx)
...
```

File: internal/controller/preflightchecks/preflight_checks.go [hunk 2]
```diff
@@ -238,3 +252,26 @@ func isPredefinedProvider(providerName string, providerType clusterctlv1.Provide
...
+	res, err := currentVersion.Compare(provider.GetSpec().Version)
+	if err != nil {
+		return false, err
+	}
+
+	// we need to delete installed components if versions are different
+	versionChanged := res != 0
+
+	return versionChanged, nil
+}
```
Comment (damdo) [inline internal/controller/preflightchecks/preflight_checks.go:277]: I know this is just codemotion, but maybe we could make this even more useful and if
`currentVersion.Compare != 0` perform a [deep.Equal()](https://pkg.go.dev/github.com/go-test/deep#Equal) and return the `[]string` with the diff between the two or else return `[]string{}, nil`.
This way we could log what is the detected difference and provide useful info to the user about it.

**Pull request closed without merging**